src/include/mpir_hw_topo.h

```diff
@@ -0,0 +1,189 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#ifndef MPIR_HW_TOPO_H_INCLUDED
+#define MPIR_HW_TOPO_H_INCLUDED
+
+/*
+ * Tracks hwloc_obj_type_t do not change ordering
+ */
+typedef enum {
+    MPIR_HWLOC_OBJ_MACHINE,
+    MPIR_HWLOC_OBJ_PACKAGE,
+    MPIR_HWLOC_OBJ_CORE,
+    MPIR_HWLOC_OBJ_PU,
+    MPIR_HWLOC_OBJ_L1CACHE,
+    MPIR_HWLOC_OBJ_L2CACHE,
+    MPIR_HWLOC_OBJ_L3CACHE,
+    MPIR_HWLOC_OBJ_L4CACHE,
+    MPIR_HWLOC_OBJ_L5CACHE,
+    MPIR_HWLOC_OBJ_L1ICACHE,
+    MPIR_HWLOC_OBJ_L2ICACHE,
+    MPIR_HWLOC_OBJ_L3ICACHE,
+    MPIR_HWLOC_OBJ_GROUP,
+    MPIR_HWLOC_OBJ_NUMANODE,
+    MPIR_HWLOC_OBJ_BRIDGE,
+    MPIR_HWLOC_OBJ_PCI_DEVICE,
+    MPIR_HWLOC_OBJ_OS_DEVICE,
+    MPIR_HWLOC_OBJ_MISC,
+    MPIR_HWLOC_OBJ_TYPE_MAX,
+    MPIR_MPICH_OBJ_DDR,         /* MPICH object for DDR memory */
+    MPIR_MPICH_OBJ_MCDRAM,      /* MPICH object for MCDRAM memory */
+    MPIR_MPICH_OBJ_DEFAULT = MPIR_MPICH_OBJ_DDR
+} MPIR_Hwloc_obj_type;
+
+/*
+ * Tracks hwloc_membind_policy_t do not change
+ */
+typedef enum {
+    MPIR_HWLOC_MEMBIND_DEFAULT = 0,
+    MPIR_HWLOC_MEMBIND_FIRSTTOUCH = 1,
+    MPIR_HWLOC_MEMBIND_BIND = 2,
+    MPIR_HWLOC_MEMBIND_INTERLEAVE = 3,
+    MPIR_HWLOC_MEMBIND_NEXTTOUCH = 4,
+    MPIR_HWLOC_MEMBIND_MIXED = -1
+} MPIR_Hwloc_membind_policy;
+
+/*
+ * Tracks hwloc_membind_flags_t do not change
+ */
+typedef enum {
+    MPIR_HWLOC_MEMBIND_PROCESS = (1 << 0),
+    MPIR_HWLOC_MEMBIND_THREAD = (1 << 1),
+    MPIR_HWLOC_MEMBIND_STRICT = (1 << 2),
+    MPIR_HWLOC_MEMBIND_MIGRATE = (1 << 3),
+    MPIR_HWLOC_MEMBIND_NOCPUBIND = (1 << 4),
+    MPIR_HWLOC_MEMBIND_BYNODESET = (1 << 5)
+} MPIR_Hwloc_membind_flags;
+
+/*
+ * MPICH specific definitions
+ */
+typedef enum {
+    MPIR_MPICH_NETLOC_OBJ_NETWORK,
+    MPIR_MPICH_NETLOC_OBJ_TOPOLOGY,
+    MPIR_MPICH_NETLOC_OBJ_NODE
+} MPIR_Netloc_obj_class;
+
+/*
+ * Tracks netloc_network_type_t do not change
+ */
+typedef enum {
+    MPIR_NETLOC_NETWORK_TYPE_ETHERNET = 1,
+    MPIR_NETLOC_NETWORK_TYPE_INFINIBAND = 2,
+    MPIR_NETLOC_NETWORK_TYPE_INVALID = 3
+} MPIR_Netloc_network_type;
```

Comment:
How about `MPIR_NETWORK_TYPE__*`?

Reply:
Actually after looking better at the code this type of information can be hidden in `topo_util.c` since it is needed only by `netloc` itself which will not be exposed directly here. What we probably want to expose to `MPIR` is just something like:

```
typedef enum {
    MPIR_NETWORK_QUERY_BY__SWITH_LEVEL,
    MPIR_NETWORK_QUERY_BY__MIN_SIZE,
    MPIR_NETWORK_QUERY_BY__MIN_MEMSIZE,
    MPIR_NETWORK_QUERY_BY__TORUS_DIMENSION
} MPIR_Network_query_by;

int (*get_covering_netobj) (MPIR_Network_query_by key, int val, int *out);
```

Where `val` can be either `switch_level`, `subcomm_min_size`, `min_mem_size` or `dimension`, respectively. Am I interpreting this correctly?